### ISSUE TITLE
segfault with exit in (?{ ... })

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2272]
Migrated from [rt.perl.org#3590](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3590) (status was 'resolved')

Searchable as RT3590$

**Comment by p5pRT** [comment]
### From jfriedl@yahoo-inc.com

#### Created by jfriedl@yahoo-inc.com



Hiho\,
Running this program with 5\.6\.0 segfaults during the exit\(\) call\. I
originally discovered the problem with a backtick \`program\` call instead of
an exit\(\)\, but this test script is smaller\.

&nbsp;    \#\!/usr/local/bin/perl \-w
&nbsp;    use strict;
&nbsp;    use re 'eval';
&nbsp;    use re 'debug';

&nbsp;    sub size\(\) \{
&nbsp;        print "BAILING\\n";
&nbsp;        exit\(0\);
&nbsp;    \}

&nbsp;    $\_ = "ABC";
&nbsp;    print "STRING&#8203;: $\_\\n";
&nbsp;    my $open;

&nbsp;    "ABC" =~ m\{
&nbsp;        A
&nbsp;        B
&nbsp;        \(?\{ size\(\) \}\)
&nbsp;        C
&nbsp;    \}x;


Oddly enough\, the print and the declaration of $open are required to trigger
the bug \-\- removing/moving either causes the segfault to not happen\.

Here's a gdb session&#8203;:

&nbsp;    gdb \./perl
&nbsp;    GNU gdb 4\.18
&nbsp;    Copyright 1998 Free Software Foundation\, Inc\.
&nbsp;    GDB is free software\, covered by the GNU General Public License\, and you are
&nbsp;    welcome to change it and/or distribute copies of it under certain conditions\.
&nbsp;    Type "show copying" to see the conditions\.
&nbsp;    There is absolutely no warranty for GDB\.  Type "show warranty" for details\.
&nbsp;    This GDB was configured as "i386\-mandrake\-linux"\.\.\.
&nbsp;    \(gdb\) run /tmp/foo2
&nbsp;    Starting program&#8203;: /usr/src/perl\-5\.6\.0/\./perl /tmp/foo2
&nbsp;    Compiling REx \`
&nbsp;    A
&nbsp;    B
&nbsp;    \(?\{ size\(\) \}\)
&nbsp;    C
&nbsp;    '
&nbsp;    size 7 first at 1
&nbsp;       1&#8203;: EXACT \<AB>\(3\)
&nbsp;       3&#8203;: EVAL\(5\)
&nbsp;       5&#8203;: EXACT \<C>\(7\)
&nbsp;       7&#8203;: END\(0\)
&nbsp;    anchored \`ABC' at 0 \(checking anchored\) minlen 3 with eval 
&nbsp;    STRING&#8203;: ABZ
&nbsp;    Guessing start of match\, REx \`
&nbsp;    A
&nbsp;    B
&nbsp;    \(?\{ size\(\) \}\)
&nbsp;    C
&nbsp;    ' against \`ABC'\.\.\.
&nbsp;    Found anchored substr \`ABC' at offset 0\.\.\.
&nbsp;    Guessed&#8203;: match at offset 0
&nbsp;    Matching REx \`
&nbsp;    A
&nbsp;    B
&nbsp;    \(?\{ size\(\) \}\)
&nbsp;    C
&nbsp;    ' against \`ABC'
&nbsp;      Setting an EVAL scope\, savestack=17
&nbsp;       0 \<> \<ABC>             |  1&#8203;:  EXACT \<AB>
&nbsp;       2 \<AB> \<C>             |  3&#8203;:  EVAL
&nbsp;      re\_eval 0x810bf08
&nbsp;    BAILING

&nbsp;    Program received signal SIGSEGV\, Segmentation fault\.
&nbsp;    Perl\_leave\_scope \(base=0\) at scope\.c&#8203;:825
&nbsp;    825                 if \(SvREFCNT\(sv\) \<= 1 && \!SvOBJECT\(sv\)\) \{
&nbsp;    \(gdb\) print sv
&nbsp;    $1 = \(SV \*\) 0x0
&nbsp;    \(gdb\) where
&nbsp;    \#0  Perl\_leave\_scope \(base=0\) at scope\.c&#8203;:825
&nbsp;    \#1  0x80aff65 in Perl\_pop\_scope \(\) at scope\.c&#8203;:144
&nbsp;    \#2  0x805ee92 in S\_my\_exit\_jump \(\) at perl\.c&#8203;:3672
&nbsp;    \#3  0x805ecf9 in Perl\_my\_exit \(status=0\) at perl\.c&#8203;:3622
&nbsp;    \#4  0x80b731d in Perl\_pp\_exit \(\) at pp\_ctl\.c&#8203;:2472
&nbsp;    \#5  0x80949ed in Perl\_runops\_standard \(\) at run\.c&#8203;:27
&nbsp;    \#6  0x401cc91a in S\_regmatch \(prog=0x810b4c4\) at re\_exec\.c&#8203;:2307
&nbsp;    \#7  0x401cabe9 in S\_regtry \(prog=0x810b488\, startpos=0x810b3e8 "ABC"\) at re\_exec\.c&#8203;:1715
&nbsp;    \#8  0x401ca4ed in my\_regexec \(prog=0x810b488\, stringarg=0x810b3e8 "ABC"\, strend=0x810b3eb ""\, 
&nbsp;    strbeg=0x810b3e8 "ABC"\, minend=0\, sv=0x810a3b4\, data=0x0\, flags=2\) at re\_exec\.c&#8203;:1509
&nbsp;    \#9  0x8096cfd in Perl\_pp\_match \(\) at pp\_hot\.c&#8203;:1027
&nbsp;    \#10 0x80949ed in Perl\_runops\_standard \(\) at run\.c&#8203;:27
&nbsp;    \#11 0x805bd4e in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:1401
&nbsp;    \#12 0x805bad8 in perl\_run \(my\_perl=0x80f0858\) at perl\.c&#8203;:1325
&nbsp;    \#13 0x8059903 in main \(argc=2\, argv=0xbffff374\, env=0xbffff380\) at perlmain\.c&#8203;:52
&nbsp;    \(gdb\) 



\(not knowing anything about perl internals\, I just issued the following
queries hoping that they'd be helpful\)



&nbsp;    \(gdb\) print PL\_savestack\_ix 
&nbsp;    3
&nbsp;    \(gdb\) print PL\_savestack@&#8203;5
&nbsp;    \{0x80fda70\, 0x80fb3e8\, 0x0\, 0x0\, 0x80f47b0\}
&nbsp;    \(gdb\) print \*PL\_savestack@&#8203;5
&nbsp;    \{\{any\_ptr = 0xffffffff\, any\_i32 = \-1\, any\_iv = \-1\, any\_long = \-1\, any\_dptr = 0xffffffff\, 
&nbsp;    any\_dxptr = 0xffffffff\}\, \{any\_ptr = 0x80ef570\, any\_i32 = 135198064\, any\_iv = 135198064\, any\_long = 135198064\, 
&nbsp;    any\_dptr = 0x80ef570 \<PL\_tmps\_floor>\, any\_dxptr = 0x80ef570 \<PL\_tmps\_floor>\}\, \{any\_ptr = 0x4\, any\_i32 = 4\, 
&nbsp;    any\_iv = 4\, any\_long = 4\, any\_dptr = 0x4\, any\_dxptr = 0x4\}\, \{any\_ptr = 0x4\, any\_i32 = 4\, any\_iv = 4\, 
&nbsp;    any\_long = 4\, any\_dptr = 0x4\, any\_dxptr = 0x4\}\, \{any\_ptr = 0x12\, any\_i32 = 18\, any\_iv = 18\, any\_long = 18\, 
&nbsp;    any\_dptr = 0x12\, any\_dxptr = 0x12\}\}

&nbsp;    \(gdb\) print PL\_savestack\[3\]\.any\_long
&nbsp;    4
&nbsp;    \(gdb\) print PL\_curpad@&#8203;10
&nbsp;    \{0x810b6e0\, 0x0\, 0x0\, 0x0\, 0x0\, 0x7b\, 0xbffff374\, 0x0\, 0x80f7c7c\, 0x80f0ae4\}
&nbsp;    \(gdb\) print \*\*PL\_curpad   
&nbsp;    \{sv\_any = 0x0\, sv\_refcnt = 2147483635\, sv\_flags = 8388608\}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by jfriedl at Sat Jul 29 20:09:33 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.15, archname=i686-linux
    uname='linux fummy.dsl.yahoo.com 2.2.16 #6 smp sun jul 23 11:26:16 pdt 2000 i686 unknown '
    config_args='-ds -e -A optimize=-g'
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2 -g', gccversion=pgcc-2.91.66 19990314 (egcs-1.1.2 release)
    cppflags='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.1.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /home/jfriedl/lib/perl
    /home/jfriedl/lib/perl/yahoo
    /usr/local/lib/perl5/5.6.0/i686-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/jfriedl
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/local/pgsql/lib:/home/jfriedl/src/rvplayer5.0
    LOGDIR (unset)
    PATH=/home/jfriedl/bin:/home/jfriedl/common/bin:/usr/local/gcc-2.95.2/bin:.:/usr/local/pgsql/bin:/usr/local/bin:/usr/X11R6/bin:/bin:/usr/bin:/usr/sbin:/sbin:/home/jfriedl/src/rvplayer5.0
    PERLLIB=/home/jfriedl/lib/perl:/home/jfriedl/lib/perl/yahoo
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Jeffrey Friedl writes&#8203;:
> Running this program with 5\.6\.0 segfaults during the exit\(\) call\.

This same script on NT segfaults also from an invalid pointer read at
scope\.c&#8203;:825\.  Maybe this Purify traceback will help&#8203;:

&nbsp;    Perl\_leave\_scope \[scope\.c&#8203;:825\]
&nbsp;            ptr = \(void\*\)&PL\_curpad\[SSPOPLONG\];
&nbsp;            sv = \*\(SV\*\*\)ptr;
&nbsp;            /\* Can clear pad variable in place? \*/
 =>         if \(SvREFCNT\(sv\) \<= 1 && \!SvOBJECT\(sv\)\) \{
&nbsp;            if \(SvTHINKFIRST\(sv\)\)
&nbsp;                sv\_force\_normal\(sv\);
&nbsp;            if \(SvMAGICAL\(sv\)\)
&nbsp;    Perl\_pop\_scope \[scope\.c&#8203;:144\]
&nbsp;    \{
&nbsp;        dTHR;
&nbsp;        I32 oldsave = PL\_scopestack\[\-\-PL\_scopestack\_ix\];
 =>     LEAVE\_SCOPE\(oldsave\);
&nbsp;    \}
&nbsp;    
&nbsp;    void
&nbsp;    S\_my\_exit\_jump \[perl\.c&#8203;:3672\]
&nbsp;            if \(cxstack\_ix > 0\)
&nbsp;                dounwind\(0\);
&nbsp;            POPBLOCK\(cx\,PL\_curpm\);
&nbsp;     =>     LEAVE;
&nbsp;            \}
&nbsp;        
&nbsp;            JMPENV\_JUMP\(2\);
&nbsp;    Perl\_my\_exit   \[perl\.c&#8203;:3622\]
&nbsp;            STATUS\_NATIVE\_SET\(status\);
&nbsp;            break;
&nbsp;            \}
&nbsp;     =>     my\_exit\_jump\(\);
&nbsp;        \}
&nbsp;        
&nbsp;        void
&nbsp;    Perl\_pp\_exit   \[pp\_ctl\.c&#8203;:2472\]
&nbsp;        \#endif
&nbsp;            \}
&nbsp;            PL\_exit\_flags |= PERL\_EXIT\_EXPECTED;
&nbsp;     =>     my\_exit\(anum\);
&nbsp;            PUSHs\(&PL\_sv\_undef\);
&nbsp;            RETURN;
&nbsp;        \}
&nbsp;    Perl\_runops\_debug \[run\.c&#8203;:56\]
&nbsp;    S\_regmatch     \[re\_exec\.c&#8203;:2307\]
&nbsp;    S\_regtry       \[re\_exec\.c&#8203;:1715\]
&nbsp;    my\_regexec     \[re\_exec\.c&#8203;:1509\]
&nbsp;    Perl\_pp\_match  \[pp\_hot\.c&#8203;:1027\]

====================================================
Mark Leighton Fisher    Thomson Consumer Electronics
fisherm@&#8203;tce\.com         Indianapolis\, IN\, USA
"Display some adaptability\."  \-\- Doug Shaftoe\, \_Cryptonomicon\_

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
